site/src/pages/WorkspacePage/WorkspaceActions/DownloadLogsDialog.tsx

```diff
@@ -28,70 +36,124 @@ type DownloadableFile = {
+  onClose,
   download = saveAs,
-  ...dialogProps
 }) => {
   const theme = useTheme();
-  const agents = selectAgents(workspace);
-  const agentLogResults = useQueries({
-    queries: agents.map((a) => ({
-      ...agentLogs(workspace.id, a.id),
-      enabled: dialogProps.open,
-    })),
-  });
+
   const buildLogsQuery = useQuery({
     ...buildLogs(workspace),
-    enabled: dialogProps.open,
+    enabled: open,
   });
-  const downloadableFiles: DownloadableFile[] = useMemo(() => {
-    const files: DownloadableFile[] = [
-      {
-        name: `${workspace.name}-build-logs.txt`,
-        blob: buildLogsQuery.data
-          ? new Blob([buildLogsQuery.data.map((l) => l.output).join("\n")], {
-              type: "text/plain",
-            })
-          : undefined,
-      },
+  const buildLogsFile = useMemo<DownloadableFile>(() => {
+    return {
+      name: `${workspace.name}-build-logs.txt`,
+      blob: buildLogsQuery.data
+        ? new Blob([buildLogsQuery.data.map((l) => l.output).join("\n")], {
+            type: "text/plain",
+          })
+        : undefined,
+    };
+  }, [workspace.name, buildLogsQuery.data]);
+
+  // This is clunky, but we have to memoize in two steps to make sure that we
+  // don't accidentally break the memo cache every render. We can't tuck
+  // everything into a single memo call, because we need to set up React Query
+  // state between processing the agents, but we can't violate rules of hooks by
+  // putting hooks inside of hooks
+  type AgentInfo = Readonly<{
+    agents: readonly WorkspaceAgent[];
+    logOptionsArray: readonly UseQueryOptions<readonly WorkspaceAgentLog[]>[];
+  }>;
+
+  const { agents, logOptionsArray } = useMemo<AgentInfo>(() => {
+    const allAgents = workspace.latest_build.resources.flatMap(
+      (resource) => resource.agents ?? [],
+    );
+
+    // Can't use the "new Set()" trick because we're not dealing with primitives
+    const uniqueAgents = [
+      ...new Map(allAgents.map((agent) => [agent.id, agent])).values(),
     ];
 
+    return {
+      agents: uniqueAgents,
+      logOptionsArray: uniqueAgents.map((agent) => {
+        return {
+          ...agentLogs(workspace.id, agent.id),
+          enabled: open,
+        };
+      }),
+    };
+  }, [workspace, open]);
+
+  const agentLogQueries = useQueries({ queries: logOptionsArray });
+  const allFiles = useMemo<readonly DownloadableFile[]>(() => {
+    const files: DownloadableFile[] = [buildLogsFile];
+
     agents.forEach((a, i) => {
       const name = `${a.name}-logs.txt`;
-      const logs = agentLogResults[i].data;
-      const txt = logs?.map((l) => l.output).join("\n");
+      const txt = agentLogQueries[i]?.data?.map((l) => l.output).join("\n");
+
       let blob: Blob | undefined;
       if (txt) {
         blob = new Blob([txt], { type: "text/plain" });
       }
+
       files.push({ name, blob });
     });
 
     return files;
-  }, [agentLogResults, agents, buildLogsQuery.data, workspace.name]);
```

Comment:
This was a problem with the previous logic: because `agents` was getting redefined every render, we were also invalidating the `useMemo` dependency array every render